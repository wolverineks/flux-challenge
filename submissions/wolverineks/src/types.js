// Types.js
// @flow

import type {Store as ReduxStore, Dispatch as ReduxDispatch} from 'redux'

import type {State as SithDashboardState} from './SithDashboard/types'

export type State = {sithDashboard: SithDashboardState}
export type Action = {type: string, data: any}

export type Store = ReduxStore<State, Action>
export type GetState = () => State
export type Dispatch = ReduxDispatch<Action>
