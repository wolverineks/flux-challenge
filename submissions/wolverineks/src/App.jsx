// @flow

import React from 'react'
import Websocket from 'react-websocket'
import {compose} from 'ramda'

import {Siths_, Planet_, List_} from './state/connectors'
import Layout, {Sith, Header, Body, PlanetIndicator, ScrollableList, List, Item, ScrollButtons, ScrollButton} from './ui/'

import {name, homeworld} from './ui/utils'

const PLANET_URL = 'ws://localhost:4000'

const SITH_URL = 'http://localhost:3000/dark-jedis/'
import sithActions from './state/actions/siths'

const App = () => <Layout>

  <Planet_>{({planet, planetReceived, dispatch}) => <div>
    <Websocket url={PLANET_URL} onMessage={(serverPlanet: string) => compose(planetReceived, JSON.parse)(serverPlanet)}/>

    <Header>
      <PlanetIndicator planet={planet} />
    </Header>

    <Body>
      <Siths_>{({siths}) =>
        <List_>{({list, scrolled}) =>
          <ScrollableList>

            <List>
              <Item><Sith name={name(siths[list[0]]) || ''} homeworld={homeworld(siths[list[0]]) || {}} /></Item>
              <Item><Sith name={name(siths[list[1]]) || ''} homeworld={homeworld(siths[list[1]]) || {}} /></Item>
              <Item><Sith name={name(siths[list[2]]) || ''} homeworld={homeworld(siths[list[2]]) || {}} /></Item>
              <Item><Sith name={name(siths[list[3]]) || ''} homeworld={homeworld(siths[list[3]]) || {}} /></Item>
              <Item><Sith name={name(siths[list[4]]) || ''} homeworld={homeworld(siths[list[4]]) || {}} /></Item>
            </List>

            <ScrollButtons>
              <ScrollButton direction={'UP'} onClick={scrolled} />
              <ScrollButton direction={'DOWN'} onClick={scrolled} />
            </ScrollButtons>

          </ScrollableList>}
        </List_>
      }</Siths_>
    </Body>

    <button onClick={() =>
      fetch(`${SITH_URL}${3616}`)
        .then((response) => response.json())
        .then((sith) => dispatch(sithActions.sithReceived(sith)))}>Start</button>
  </div>}</Planet_>

</Layout>

export default App
