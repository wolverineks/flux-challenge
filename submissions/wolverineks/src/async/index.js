// @flow

import {combinePixies} from 'redux-pixies'

import sithPixie from './sithPixie'
import sithsPixie from './sithsPixie'

export {
  sithPixie,
  sithsPixie
}

export default combinePixies({sithsPixie})
