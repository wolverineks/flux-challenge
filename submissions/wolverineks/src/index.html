<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <link rel="stylesheet" href="./styles.css">
  <title>Flux Challenge</title>
</head>

<body>
  <div id="root"></div>
</body>

</html>
