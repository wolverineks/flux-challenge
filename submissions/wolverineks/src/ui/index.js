// @flow

import type {Children} from '../state/types'
import {Layout, Header, Body, PlanetIndicator, Sith, ScrollableList, ScrollButtons, ScrollButton, List, Item} from './components'
import classNames from './styles'

export type {Children}
export {
  Layout,
  Header,
  Body,
  PlanetIndicator,
  Sith,
  ScrollableList,
  ScrollButtons,
  ScrollButton,
  List,
  Item,
  classNames
}
export default Layout
