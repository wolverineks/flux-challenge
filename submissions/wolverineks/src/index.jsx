// @flow

import React from 'react'
import ReactDOM from 'react-dom'
import {Provider} from 'react-redux'
import {attachPixie} from 'redux-pixies'

import createStore from './state/store'
import App from './App.jsx'
import appPixie from './async'

const store = createStore({})

attachPixie(store, appPixie)

ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
)
