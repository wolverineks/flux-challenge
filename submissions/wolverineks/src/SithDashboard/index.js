// SithDashboard/index.js
// @flow

import selectors from './selectors'
import reducer from './reducer'

import Main from './SithDashboard.jsx'
import connector from './connector'

export {
  Main,
  reducer,
  selectors,
}

export default connector
