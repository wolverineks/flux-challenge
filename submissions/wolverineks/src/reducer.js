// reducer.js
// @flow

import {combineReducers} from 'redux'

import {reducer} from './SithDashboard'

export default combineReducers({
  sithDashboard: reducer
})
