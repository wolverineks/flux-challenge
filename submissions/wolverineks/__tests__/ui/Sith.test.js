// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import {Sith, classNames} from '../../src/ui/'

describe('Sith', () => {
  describe('When has name', () => {
    describe('When has planet', () => {
      test('Sith', () => {
        const className = classNames.sith
        const name = 'Skere Kaan'
        const planet = {id: 1, name: 'Coruscant'}
        const lastUpdated = new Date(1510591530092)
        const component = renderer.create(<Sith name={name} homeworld={planet} className={className} lastUpdated={lastUpdated} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('When lacks planet', () => {
      test('Sith', () => {
        const className = classNames.sith
        const name = 'Skere Kaan'
        const planet = undefined
        const lastUpdated = new Date(1510591530092)
        const component = renderer.create(<Sith name={name} homeworld={planet} className={className} lastUpdated={lastUpdated} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
  })

  describe('When lacks name', () => {
    describe('When has planet', () => {
      test('Sith', () => {
        const className = classNames.sith
        const name = undefined
        const planet = {id: 1, name: 'Coruscant'}
        const lastUpdated = new Date(1510591530092)
        const component = renderer.create(<Sith name={name} homeworld={planet} className={className} lastUpdated={lastUpdated} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('When lacks planet', () => {
      test('Sith', () => {
        const className = classNames.sith
        const name = undefined
        const planet = undefined
        const lastUpdated = undefined
        const component = renderer.create(<Sith name={name} homeworld={planet} className={className} lastUpdate={lastUpdated} />)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
  })
})
