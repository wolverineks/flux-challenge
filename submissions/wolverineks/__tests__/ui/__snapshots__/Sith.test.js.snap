// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sith When has name When has planet Sith 1`] = `
<div>
  <h3>
    Skere Kaan
  </h3>
  <h6>
    Homeworld: Coruscant
  </h6>
</div>
`;

exports[`Sith When has name When lacks planet Sith 1`] = `
<div>
  <h3>
    Skere Kaan
  </h3>
  <h6 />
  <h6>
    9 days ago
  </h6>
</div>
`;

exports[`Sith When lacks name When has planet Sith 1`] = `
<div>
  <h3 />
  <h6>
    Homeworld: Coruscant
  </h6>
</div>
`;

exports[`Sith When lacks name When lacks planet Sith 1`] = `
<div>
  <h3 />
  <h6 />
  <h6>
    a few seconds ago
  </h6>
</div>
`;
